cluster/buckets.go

```diff
@@ -0,0 +1,266 @@
+	if err != nil {
+		return []TenantBucketInfo{}, err
+	}
+
+	var buckets []minio.BucketInfo
+	buckets, err = minioClient.ListBuckets()
+	if err != nil {
+		return []TenantBucketInfo{}, tagErrorAsMinio(err)
+	}
+
+	var (
+		accessType  BucketAccess
+		bucketInfos []TenantBucketInfo
+	)
+	for _, bucket := range buckets {
+		accessType, err = GetBucketAccess(minioClient, bucket.Name)
+		if err != nil {
+			return []TenantBucketInfo{}, tagErrorAsMinio(err)
+		}
+		bucketInfos = append(bucketInfos, TenantBucketInfo{Name: bucket.Name, Access: accessType})
+	}
+
+	return bucketInfos, nil
+}
+
+// Deletes a bucket in the given tenant's MinIO
+func DeleteBucket(tenantShortname, bucket string) error {
+	// Get tenant specific MinIO client
+	minioClient, err := newTenantMinioClient(nil, tenantShortname)
+	if err != nil {
+		return err
+	}
+
+	return minioClient.RemoveBucket(bucket)
+}
+
+func registerBucketForTenant(ctx *Context, bucketName string, tenantID *uuid.UUID) error {
+	tx, err := ctx.MainTx()
+	if err != nil {
+		return err
+	}
+	// create the bucket registry
+	query :=
+		`INSERT INTO
+			buckets ("name","tenant_id")
+		VALUES
+			($1, $2)
+		ON CONFLICT DO NOTHING`
+
+	if _, err = tx.Exec(query, bucketName, tenantID); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func unregisterBucketForTenant(ctx *Context, bucketName string, tenantID *uuid.UUID) error {
+	tx, err := ctx.MainTx()
+	if err != nil {
+		return err
+	}
+	// delete the bucket registry
+	query :=
+		`DELETE FROM
+			buckets 
+		WHERE name=$1 AND tenant_id=$2`
+
+	if _, err = tx.Exec(query, bucketName, tenantID); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+type BucketToService struct {
+	Bucket      string
+	Service     string
+	ServicePort int32
+}
+
+type BucketToServiceResult struct {
+	BucketToService *BucketToService
+	Error           error
+}
+
+// streamBucketToTenantServices returns a channel that will receive a list of buckets and the domain tenant service
+// they resolve to.
+// This function uses a channel because there may be hundreds of thousands of buckets and we don't want to pre-alloc
+// all that information on memory.
+func streamBucketToTenantServices() chan *BucketToServiceResult {
+	ch := make(chan *BucketToServiceResult)
+	go func() {
+		defer close(ch)
+		query :=
+			`SELECT 
+				b.name, tsg.service_name, tsg.port
+			FROM 
+				buckets b 
+				LEFT JOIN tenants_storage_groups tsg ON b.tenant_id = tsg.tenant_id
+			ORDER BY b.name ASC`
+
+		// no context? straight to db
+		rows, err := GetInstance().Db.Query(query)
+		if err != nil {
+			ch <- &BucketToServiceResult{Error: err}
+			return
+		}
+		defer rows.Close()
+
+		for rows.Next() {
+			// Save the resulted query on the User struct
+			b2s := BucketToService{}
+			err = rows.Scan(&b2s.Bucket, &b2s.Service, &b2s.ServicePort)
+			if err != nil {
+				ch <- &BucketToServiceResult{Error: err}
+				return
+			}
+			ch <- &BucketToServiceResult{BucketToService: &b2s}
+		}
```

Comment:
they are being returned for the caller to manage the error, since this is a streaming function it's up to the caller to decide what to do

Reply:
`http://go-database-sql.org/retrieving.html`
We can do it similar to how it is here 
typo: It should be `rows.Err()` instead of rows.Close()
```
err = rows.Err()
if err != nil {
	...
}
```
At the end because scan could probably not trigger ab error.